Filen Backend

Enduriel opened the pull request:
#### What is the purpose of this change?

Add Support for [Filen](https://filen.io/)

#### Was the change discussed in an issue or in the forum before?

Discussed in issue #6728

#### Known (potential) Issues

- The backend still fails a single test `TestBisyncRemoteRemote/extended_filenames`, I've not been able to figure out why, so I'd appreciate some help with debugging this. [Here's](https://github.com/user-attachments/files/20026962/2025-05-04-101750.zip) a zip of the report of my latest run.
- I'm not 100% clear on whether I can integrate with rclone's standard multithreading approach. The way filen uploads work is that uploads are always made in 1 MiB chunks, this didn't seem to integrate well with the multithreading assumptions rclone makes (especially because this means non-multithreaded speeds are awful), so there's currently a somewhat different multithreading system which can be configured as specified in the documentation.

#### Other Questions
- In the longer term, we're considering transitioning to a unified rust SDK for all our products, I saw there was some discussion on whether it'd be feasible to use an underlying rust library for a backend [here](https://forum.rclone.org/t/would-calling-gpl-licensed-rust-be-acceptable/48365/3), and I just wanted to clarify if you guys think it would be possible as a long term goal to potential use that.

#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-new-feature-or-bug-fix).
- [x] I have added tests for all changes in this PR if appropriate.
- [x] I have added documentation for the changes if appropriate.
- [x] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [ ] I'm done, this Pull Request is ready for review :-)

ashutoshsaboo commented:
@rclone team - can someone please help and get this PR merged? Been an active Filen user, and would love to see their rclone functionality go GA 🚀 . Have already used their rclone beta build, and it works fab!

danieletorelli commented:
> @rclone team - can someone please help and get this PR merged? Been an active Filen user, and would love to see their rclone functionality go GA 🚀 . Have already used their rclone beta build, and it works fab!

My +1 on this, I'm also using the beta from a while now and it works

ncw commented:
Hi @Enduriel sorry for the delay looking at this. I wanted to get the very delayed v1.70 out the door first.

> * The backend still fails a single test `TestBisyncRemoteRemote/extended_filenames`, I've not been able to figure out why, so I'd appreciate some help with debugging this. [Here's](https://github.com/user-attachments/files/20026962/2025-05-04-101750.zip) a zip of the report of my latest run.

It sounds like @erusc has got to the bottom of this. We could just decide to ignore this as it is very unlikely to affect users. Or we could adjust the encoding to cope with those characters.

> * I'm not 100% clear on whether I can integrate with rclone's standard multithreading approach. The way filen uploads work is that uploads are always made in 1 MiB chunks, this didn't seem to integrate well with the multithreading assumptions rclone makes (especially because this means non-multithreaded speeds are awful), so there's currently a somewhat different multithreading system which can be configured as specified in the documentation.

Fixed size 1MiB chunks should work perfectly with rclone's multithread approach if you implement `OpenChunkWriter`.

https://github.com/rclone/rclone/blob/5aa98110845d436a85d7c6db81b6dde40f0edac9/fs/features.go#L187-L192

Check out the s3 backend if you want to see an implementation.

> * I can't figure out how to satisfy this linter for the docs section, it's all autogenerated.

That is a bug which should be fixed if you rebase!

> #### Other Questions
> * In the longer term, we're considering transitioning to a unified rust SDK for all our products, I saw there was some discussion on whether it'd be feasible to use an underlying rust library for a backend [here](https://forum.rclone.org/t/would-calling-gpl-licensed-rust-be-acceptable/48365/3), and I just wanted to clarify if you guys think it would be possible as a long term goal to potential use that.

We don't use CGO so we can't link to C libraries. This makes it really easy to generate single binaries which work everywhere. So we can't use the obvious approach - a shared C library. There are other approaches though like compile to wasm and run with [wazero](https://wazero.io/) or the [purego](https://github.com/ebitengine/purego) approach. We use something very like purego for running with WinFSP which is loaded at runtime.

I haven't reviewed the code yet, but I thought I'd get the discussion started.

Enduriel commented:
Thanks for digging into the `extended_filenames` issue @erusc, that's enough for me to debug by. I'll try and get this figured out by end of day tomorrow.

@ncw thanks for getting back to me even with just preliminary stuff. I'll rebase and update for the linter issue and hopefully fix the encoding issue very soon.

Regarding the multithreading, the ChunkWriter is a good point and I do think would solve half of the issues (uploading), unfortunately, without the threading I do right now, this would still leave the download speeds very slow.

With the current Filen API, the way it works is that each chunk is a distinct request on upload/download. This means that if you are sequentially downloading/uploading a file, the speed is really abysmal, especially for users far from the data centers. (FWIW I'm hoping we'll be able to pivot away from this approach in the future but if that happens I expect that to take at least a year :/).

Let me know what you think the best approach would be with this information.

Enduriel commented:
They are, is there a way to hint to rclone what the minimum chunk size is for downloads? If that is the case then yeah I think that would work.

Dwynr commented:
> They are, is there a way to hint to rclone what the minimum chunk size is for downloads? If that is the case then yeah I think that would work.
> 
> ETA: Otherwise I am concerned about a lot of unnecessary downloaded chunks that overlap

@ncw sorry for pinging you.. but any further comment on that? We really want to get our backend merged into main asap :) We'll modify it accordingly as soon as we get more info.

ncw commented on backend/filen/filen.go:
```diff
@@ -0,0 +1,1084 @@
+}
+
+// File is Filen's normal file
+type File struct {
```
Could you call this `Object` please to match the other backends otherwise it is going to confuse us!

The methods should be `func (o *Object) Method()` also.

ncw commented on backend/filen/filen.go:
```diff
@@ -0,0 +1,1084 @@
+		return "", hash.ErrUnsupported
+	}
+	if file.file.Hash == "" {
+		foundFile, err := file.fs.filen.FindFile(ctx, file.fs.resolvePath(file.path))
```
Will reading the Hash take another transaction normally? Or does the Hash arrive in the directory listings?

If it always takes another transaction then you should set the `SlowHash` feature flag.

Enduriel replied:
Hash arrives in directory listings, it's part of the encrypted metadata so anytime we get file info we also get the hash

Enduriel commented:
- Will do
- Will do after I rebase & cleanup
- I'll do that for now, I guess I can PR later to fix this as well, I might try and fix this after the cleanup depending on timing.

Yes there shouldn't be any issues with the free tier.

ncw commented on backend/filen/filen.go:
```diff
@@ -0,0 +1,1084 @@
+// dirMoveContents moves the contents of srcDir to dstDir
+// used for the case where the target directory exists
+// recurses if needed
+func (f *Fs) dirMoveContents(ctx context.Context, srcDir, dstDir types.DirectoryInterface, srcPath, dstPath string) error {
```
I think rclone should do this for you.

`DirMove` is allowed to fail if the destination exists.

That works fine for `rclone mount` and for `rclone move` it will engage backup code which does what this does.

https://github.com/rclone/rclone/blob/ff9cbab5fa36fef07f9237dcef073e16b0c72de8/fs/sync/sync.go#L1401

https://github.com/rclone/rclone/blob/ff9cbab5fa36fef07f9237dcef073e16b0c72de8/fs/sync/sync.go#L1421

So I think you should remove this function and just return `fs.ErrorDirExists`

NB the help for `DirMove` says

> If destination exists then return fs.ErrorDirExists

Enduriel replied:
Well that is annoying to have missed, I will simplify this.

ncw commented on backend/filen/filen.go:
```diff
@@ -0,0 +1,1084 @@
+	}
+
+	if dstDir != nil {
+		return f.dirMoveContents(ctx, srcDirInt, dstDir, srcPath, dstPath)
```
This should just return fs.ErrorDirExists see function comment.

ncw commented on backend/filen/filen.go:
```diff
@@ -0,0 +1,1084 @@
+)
+
+// todo PublicLinker,
+// we could technically implement ChangeNotifier, but
```
ChangeNotify is great for long running `rclone mount`. Yes rclone's implementation is very lowest common denominator - it is just use to flush the directory cache when we know a directory has changed.

ncw commented on docs/content/filen.md:
```diff
@@ -0,0 +1,265 @@
+title: "Filen"
+description: "Rclone docs for Filen"
+versionIntroduced: "TODO"
+status: Alpha
```
I'd leave out the status. If it passes all the integration tests then it is much better than Alpha :-)

ncw commented on docs/content/filen.md:
```diff
@@ -0,0 +1,265 @@
+---
+title: "Filen"
+description: "Rclone docs for Filen"
+versionIntroduced: "TODO"
```
v1.71 (cross fingers!)

ncw commented on backend/filen/filen.go:
```diff
@@ -0,0 +1,1084 @@
+		}
+	}
+
+	filen.MaxDownloadThreadsPerFile = opt.MaxDownloadThreadsPerFile
```
Are these actually used in the backend's usage? Given that rclone backends stream everything and don't read or write to files directly?

Enduriel commented on backend/filen/filen.go:
```diff
@@ -0,0 +1,1084 @@
+				Required: true,
+			},
+			{
+				Name:       "password",
```
Yes, unfortunately this is due to our encryption, the key is derived from the password, auth/login is separate however and can be blocked by 2FA (and also rate limited), so this was the only way I can think of to make this work.

Enduriel commented:
I'm working on fixing the failing test (and getting my mind blown by it), but regarding the downloads, it might be easier for you to see what I mean rather than me (badly) trying to explain it to you. See [here](https://github.com/FilenCloudDienste/filen-sdk-go/blob/ac9988ad35cd124eab26827d1aa15362c6e343c7/filen/download.go#L88)

Enduriel commented:
I've managed to isolate my failing test to this minimal snippet. I'm not sure if this is a bug in the encoder or if I'm misunderstanding how this should work but
```go
enc := (encoder.Standard | encoder.EncodeInvalidUtf8)
str := "\n"
fmt.Printf("encoder makes sense: %v\n", enc.FromStandardPath(enc.ToStandardPath(str)) == str)
```
will have `enc.FromStandardPath(enc.ToStandardPath(str)) == str` evaluate to false, which doesn't make sense to me (but again, maybe that's my flawed understanding of how it should work?).

KatyPurry commented:
Wasn't it scheduled for 1.72? What's holding it up? This is starting to take a long time.

Enduriel commented:
Thanks for getting back to me, and aiming to get this into 1.73 sounds good.

The failing test has been fixed, so the last time I ran it, the backend passed all the test suites fully.

Regarding uploads/downloads, alright, I will work on dumbing down the backend so that downloads/uploads are fully sequential per thread, the only thing that's unfortunate here is that with downloads there could be a significant amount of extra work done when downloading with overlapping chunks, which I was hoping to avoid, but if you're saying that there's no way to avoid that, then it is what it is.

KatyPurry commented:
> Looks like 1.73 will be missed too?

Why ? Not again please…

Enduriel commented:
Quick update from my side before a material update to the PR hopefully tomorrow. I've implemented ChunkWriter and removed all the backend side parallelization like we discussed.

This isn't currently an issue but I do not allow Chunk sizes that are not multiples of 1 MiB, this would be very annoying to handle for this backend and basically remove all the parallelization benefits, I hope that's not a problem, it still passes all the relevant tests.

I've also fixed lacking support for SeekOption and the lack of logging unsupported mandatory options for future-proofing.

I still need to rebase and test on the latest master branch, but I suspect there won't be any issues there.

Just finished a [test](https://github.com/user-attachments/files/24532832/2026-01-09-172718.zip) confirming that the ChunkWriter stuff is working as well.

Still a couple minor tweaks, but getting there.
